2022/Q3/artifacts/PoS-tla/Staking.tla

```diff
@@ -113,104 +153,245 @@ Delegate(sender, validator, amount) ==
+          UNCHANGED <<balanceOf, posAccount, bonded, idBonds, totalDeltas>>
         ELSE
           \* transaction succeeds
           \* update the balance of 'sender'
           /\ balanceOf' = [ balanceOf EXCEPT ![sender] = @ - amount]
-          /\ delegated' = [ n \in 1..UnbondingLength, user \in UserAddrs, val \in ValidatorAddrs |-> 
-                     IF n >= PipelineLength /\ user = sender /\ val = validator
-                     THEN delegated[n, user, val] + amount
-                     ELSE delegated[n, user, val]
-                    ]
-          /\ bonded' = [ n \in 1..UnbondingLength, user \in ValidatorAddrs |-> 
-                     IF n >= PipelineLength /\ user = sender
-                     THEN bonded[n, user] + amount
-                     ELSE bonded[n, user]
-                    ]
-          /\ UNCHANGED unbonded
-
-
-\* unbond `amount` tokens from a validator
+          /\ posAccount' = posAccount + amount
+          /\ bonded' = [ bonded EXCEPT ![sender, validator] = @ \union {[ id |-> idBonds,
+                                                                          epoch |-> epoch + PipelineLength,
+                                                                          amount |-> amount ]}]
+          /\ idBonds' = idBonds + 1
+          \* updates totalDeltas from PipelineLength to UnbondingLength 
+          /\ totalDeltas' = [ n \in 0..2*UnbondingLength, val \in ValidatorAddrs |-> totalDeltas[n, val] + 
+                              IF n >= UnbondingLength + PipelineLength /\ val = validator
+                              THEN amount
+                              ELSE 0
+                            ]
+
+\* Follows the specification: it counts all bonds, even those that have not been materialized yet.
+\* PAY ATTENTION TO IT
+\* @type: (ADDR, ADDR) => Int;
+TotalBonds(sender, validator) == LET 
+                                 \* @type: (Int, BOND) => Int;
+                                 F(total, bond) == total + bond.amount
+                                 IN ApaFoldSet(F, 0, bonded[sender, validator])
+
+\* @type: (Set(BOND), Int, Int) => [ remaining: Int, set: Set(UNBOND), id: Int ];
+ComputedUnbonds(setBonds, totalAmount, e) == LET 
+                                         \* @type: ([ remaining: Int, set: Set(UNBOND), id: Int ], BOND) => [ remaining: Int, set: Set(UNBOND), id: Int ];
+                                         F(record, bond) == 
+                                         IF record.remaining > 0
+                                         THEN 
+                                          LET min == Min(record.remaining, bond.amount) 
+                                          IN
+                                          [ remaining |-> record.remaining - min,
+                                            set |-> record.set \union {[id |-> record.id, amount |-> min, start |-> bond.epoch, end |-> e]},
+                                            id |-> record.id + 1 ]   
```

Comment:
Can you clarify the goal of incrementing the id here?

Reply:
Each unbond (the same for slashes and bonds) is uniquely identified by the id. It is just a way of transforming a set into a multiset or bag.

Reply:
Yes, but I don't get how you _know_ that `record.id + 1` is the one you're looking for... I guess this requires some synchronous interaction

Reply:
That piece of code simply creates new unbonds, each unbond gets assigned a unique id.